clarity/src/vm/types/signatures.rs

```diff
@@ -645,19 +717,38 @@ impl TypeSignature {
         }
     }
 
-    /// Canonicalize a simple type. This method will replace any simple types
-    /// from previous epochs with the appropriate types for the current epoch.
-    /// Note that complex types (`optional`, `list`, `response`, `tuple`) should
-    /// be recursed by the caller.
-    pub fn canonicalize_simple_type(&self, epoch: &StacksEpochId) -> TypeSignature {
+    /// Canonicalize a type.
+    /// This method will convert types from previous epochs with the appropriate
+    /// types for the specified epoch.
+    pub fn canonicalize(&self, epoch: &StacksEpochId) -> TypeSignature {
         match epoch {
-            StacksEpochId::Epoch21 => self.canonicalize_simple_type_v2_1(),
+            StacksEpochId::Epoch21 => self.canonicalize_v2_1(),
             _ => self.clone(),
         }
     }
 
-    pub fn canonicalize_simple_type_v2_1(&self) -> TypeSignature {
+    pub fn canonicalize_v2_1(&self) -> TypeSignature {
         match self {
+            SequenceType(SequenceSubtype::ListType(ref list_type)) => {
+                SequenceType(SequenceSubtype::ListType(ListTypeData {
+                    max_len: list_type.max_len,
+                    entry_type: Box::new(list_type.entry_type.canonicalize_v2_1()),
+                }))
+            }
+            OptionalType(ref inner_type) => OptionalType(Box::new(inner_type.canonicalize_v2_1())),
+            ResponseType(ref inner_type) => ResponseType(Box::new((
+                inner_type.0.canonicalize_v2_1(),
+                inner_type.1.canonicalize_v2_1(),
+            ))),
+            TupleType(ref tuple_sig) => {
+                let mut canonicalized_fields = BTreeMap::new();
+                for (field_name, field_type) in tuple_sig.get_type_map() {
+                    canonicalized_fields.insert(field_name.clone(), field_type.canonicalize_v2_1());
```

Comment:
is it possible to modify in place instead? maybe implement a `get_mut_type_map`?

Reply:
The rest of the cases don't require `&mut self`, mainly because I thought it was bad practice to modify `self` in the case of the simple types. That's why this method takes an immutable `self` and returns a `TypeSignature`.